src/backend/distributed/transaction/transaction_management.c

```diff
@@ -913,3 +946,212 @@ EnsurePrepareTransactionIsAllowed(void)
 					errmsg("cannot use 2PC in transactions involving "
 						   "multiple servers")));
 }
+
+
+/*
+ * CurrentTransactionPropagatedObjects returns the objects propagated in current
+ * sub-transaction or the root transaction if no sub-transaction exists.
+ */
+static HTAB *
+CurrentTransactionPropagatedObjects(void)
+{
+	if (activeSubXactContexts == NIL)
+	{
+		/* hashset in the root transaction if there is no sub-transaction */
+		return PropagatedObjectsInTx;
+	}
+
+	/* hashset in top level sub-transaction */
+	SubXactContext *state = llast(activeSubXactContexts);
+	if (state->propagatedObjects == NULL)
+	{
+		/* lazily create hashset for sub-transaction */
+		state->propagatedObjects = CreateSubtransactionPropagatedObjectsHash();
+	}
+	return state->propagatedObjects;
+}
+
+
+/*
+ * ParentTransactionPropagatedObjects returns the objects propagated in parent
+ * transaction of active sub-transaction. It returns the root transaction if
+ * no sub-transaction exists.
+ */
+static HTAB *
+ParentTransactionPropagatedObjects(void)
+{
+	if (activeSubXactContexts == NIL)
+	{
+		return PropagatedObjectsInTx;
+	}
+
+	int nestingLevel = list_length(activeSubXactContexts);
+	if (nestingLevel == 1)
```

Comment:
We should probably be a bit defensive here, given the rest of the function expects the list to be of length >= 2
```suggestion
	if (nestingLevel <= 1)
```

Reply:
This also would suggest that the check above can be removed, since an empty list (`NIL`) will return 0 as its length, being caught by this check as well and returning the same value.